### PR TITLE
DEPR: Deprecated passing arguments as positional in pd.concat

### LINKED ISSUE #41485: Deprecate non-keyword arguments for methods with `inplace`
## Background

Having many parameters with defaults makes static typing a lot harder when one of them changes the return type, as they require many* overloads.

There's still no broad agreement about whether or not to deprecate the `inplace` argument, so in the meanwhile, if we force default-arguments to be keyword-only, then we can at least set the correct return types for these methods without requiring all these overloads (as many as 38 in one case!).

## What to do

1. Choose a method from the following list:

- [ ] pandas/core/computation/eval.py:164:0, eval (skip for now)
- [x] pandas/core/frame.py:4718:4, set_axis
- [x] pandas/core/frame.py:4743:4, drop
- [x] pandas/core/frame.py:5134:4, fillna
- [x] pandas/core/frame.py:5306:4, set_index
- [x] pandas/core/frame.py:5572:4, reset_index
- [x] pandas/core/frame.py:5809:4, dropna
- [x] pandas/core/frame.py:5958:4, drop_duplicates
- [x] pandas/core/frame.py:6196:4, sort_values
- [x] pandas/core/frame.py:6267:4, sort_index
- [x] pandas/core/generic.py:6432:4, ffill
- [x] pandas/core/generic.py:6495:4, bfill
- [x] pandas/core/generic.py:6700:4, interpolate
- [x] pandas/core/generic.py:7473:4, clip
- [x] pandas/core/generic.py:9080:4, where
- [x] pandas/core/generic.py:9234:4, mask
- [x] pandas/core/indexes/base.py:1530:4, set_names
- [x] pandas/core/indexes/multi.py:810:4, set_levels
- [x] pandas/core/indexes/multi.py:997:4, set_codes
- [x] pandas/core/resample.py:831:4, interpolate
- [x] pandas/core/series.py:1279:4, reset_index
- [x] pandas/core/series.py:2028:4, drop_duplicates
- [x] pandas/core/series.py:3228:4, sort_values
- [x] pandas/core/series.py:3438:4, sort_index
- [x] pandas/core/series.py:4475:4, set_axis
- [x] pandas/core/series.py:4490:4, drop
- [x] pandas/core/series.py:4713:4, fillna
- [x] pandas/core/series.py:5064:4, dropna
- [x] read_csv
- [x] read_table
- [x] concat

2. Use `pandas.util._decorators.deprecate_nonkeyword_arguments` to issue a FutureWarning that the method's default arguments will be keyword-only in the next release.

3. Fixup any tests / internal code which uses such default arguments positionally. See [this gist](https://gist.github.com/MarcoGorelli/1e8fb9d4cfb24353b0022f03150651a0) for a helper-script which can help identify cases which need fixing, just replace `"drop"` with whatever method you're working on and then run it as `python visitor.py`.

4. Add a new test, checking that a FutureWarning is raised if this method's default arguments are used positionally.

5. Add a note to `doc/source/whatsnew/v1.3.0.rst`

See https://github.com/pandas-dev/pandas/pull/41511 for an example of how to do this.

In most of these methods, the first argument is probably fine to be kept as is, it's the ones after that (such as `axis`) which should become keyword-only.

----

*(2^{number of default types preceding the one to overload} +2)

----

It's really common to call `replace` with both `value` and `to_replace` as positional, so I think we can make an exception for that one. It'll only require 6 overloads, which I'd argue isn't too many, especially if no other method requires more than 3 or 4

---

TODO: unify the whatsnew entries

### PR DESCRIPTION
**Pull request by tegardp**
- [x] xref #41485
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/concat/test_concat.py [hunk 1]
```diff
@@ -638,3 +638,18 @@ def test_concat_multiindex_with_empty_rangeindex():
...
+    # https://github.com/pandas-dev/pandas/issues/41485
+    df = pd.DataFrame([[1, 2, 3]], index=["a"])
+    df2 = pd.DataFrame([[4, 5, 6]], index=["b"])
+
+    msg = (
+        "In a future version of pandas all arguments of concat"
+        "except for the argument 'objs' will be keyword-only"
+    )
+    with tm.assert_produces_warning(FutureWarning, match=msg):
+        result = concat([df, df2], " ")
```
Comment (MarcoGorelli) [inline pandas/tests/reshape/concat/test_concat.py:653]: can you put `0` as the second argument (or `1`) - I don't think anyone would pass " " to `axis`

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Generally looks good, thanks! Can you drop the commits from the `read_table` PR here?

File: pandas/tests/reshape/concat/test_concat.py (hunk 1 above)
Comment (tegardp) [reply pandas/tests/reshape/concat/test_concat.py]: done, thanks for the pointer :)

**Comment by jreback** [comment]
can you rebase; this appears to have a read_table change (already merged) so need to remove that

File: pandas/tests/reshape/concat/test_concat.py [hunk 2]
```diff
@@ -638,3 +638,18 @@ def test_concat_multiindex_with_empty_rangeindex():
...
+    # https://github.com/pandas-dev/pandas/issues/41485
+    df = pd.DataFrame([[1, 2, 3]], index=["a"])
+    df2 = pd.DataFrame([[4, 5, 6]], index=["b"])
+
+    msg = (
+        "In a future version of pandas all arguments of concat"
+        "except for the argument 'objs' will be keyword-only"
+    )
+    with tm.assert_produces_warning(FutureWarning, match=msg):
+        result = concat([df, df2], axis=0)
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/concat/test_concat.py:653]: axis argument should be positional

File: pandas/tests/reshape/concat/test_concat.py [hunk 3]
```diff
@@ -638,3 +638,18 @@ def test_concat_multiindex_with_empty_rangeindex():
     result = concat([df1, df2])
     expected = DataFrame([[1, 2], [np.nan, np.nan]], columns=mi)
     tm.assert_frame_equal(result, expected)
+
+
+def test_concat_posargs_deprecation(all_parsers):
+    # https://github.com/pandas-dev/pandas/issues/41485
+    df = pd.DataFrame([[1, 2, 3]], index=["a"])
+    df2 = pd.DataFrame([[4, 5, 6]], index=["b"])
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/concat/test_concat.py:646]: pandas/tests/reshape/concat/test_concat.py:645:10: PDF019 found both 'pd.DataFrame' and 'DataFrame' in the same file
pandas/tests/reshape/concat/test_concat.py:646:11: PDF019 found both 'pd.DataFrame' and 'DataFrame' in the same file

**Comment by jreback** [comment]
@tegardp if you can update

**Comment by tegardp** [comment]
> @tegardp if you can update

sorry i'm still confused about rebase workflow

**Comment by simonjayhawkins** [comment]
> > @tegardp if you can update
> 
> sorry i'm still confused about rebase workflow

that is now sorted in https://github.com/pandas-dev/pandas/pull/41718/commits/d9b371252f1e8e4ae7acc812d25eb7b5080efbf0,

comments https://github.com/pandas-dev/pandas/pull/41718#discussion_r643799672 and https://github.com/pandas-dev/pandas/pull/41718#discussion_r643811201 to be addressed. will fix test and lint failures on ci

File: pandas/tests/reshape/concat/test_concat.py [hunk 4]
```diff
@@ -638,3 +638,18 @@ def test_concat_multiindex_with_empty_rangeindex():
     result = concat([df1, df2])
     expected = DataFrame([[1, 2], [np.nan, np.nan]], columns=mi)
     tm.assert_frame_equal(result, expected)
+
+
+def test_concat_posargs_deprecation(all_parsers):
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/concat/test_concat.py:643]: and why is all_parsers here?

File: pandas/tests/reshape/concat/test_concat.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/tests/reshape/concat/test_concat.py:653]:
```suggestion
        result = concat([df, df2], 0)
```

**Comment by simonjayhawkins** [comment]
> comments [#41718 (comment)](https://github.com/pandas-dev/pandas/pull/41718#discussion_r643799672) and [#41718 (comment)](https://github.com/pandas-dev/pandas/pull/41718#discussion_r643811201) to be addressed. will fix test and lint failures on ci

tests are still failing... see https://github.com/pandas-dev/pandas/pull/41718#discussion_r643799672

**Review by jreback — APPROVED** [review]
great ping on green

**Comment by jreback** [comment]
you need to change any existing test which uses positional args to avoid the warnings.

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/pull/41718/checks?check_run_id=2747849764 one failed test

File: pandas/tests/reshape/concat/test_invalid.py [hunk 5]
```diff
@@ -28,11 +28,16 @@ def test_concat_invalid(self):
...
             "first argument must be an iterable of pandas "
             'objects, you passed an object of type "DataFrame"'
         )
-        with pytest.raises(TypeError, match=msg):
+        warn_msg = (
+            "In a future version of pandas all arguments of concat "
+            "except for the argument 'objs' will be keyword-only"
+        )
+        warning_ctx = tm.assert_produces_warning(FutureWarning, match=warn_msg)
+        with pytest.raises(TypeError, match=err_msg), warning_ctx:
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/concat/test_invalid.py:40]: when we deprecate, python will raise `TypeError: concat() takes 1 positional argument but 2 were given` before the function validates the first argument.

maybe `concat(df1, df2)` -> `concat(df1)`

**Review by simonjayhawkins — APPROVED** [review]
Thanks @tegardp and @MarcoGorelli lgtm

ci failures are codecov failed uploads.

**Comment by jreback** [comment]
thanks @tegardp

**Pull request merged**